client/asset/btc/btc.go

```diff
@@ -5474,6 +5711,75 @@ func (btc *baseWallet) BondsFeeBuffer() uint64 {
 	return 15 * bondsFeeBuffer(btc.segwit, btc.feeRateLimit()) / 10
 }
 
+// FundMultiOrder funds multiple orders in one shot. MaxLock is the maximum
+// amount that the wallet can lock for these orders. If maxLock == 0, then
+// there is no limit. An error is returned if the wallet does not have enough
+// available balance to fund each of the orders, however, if splitting is
+// not enabled and all of the orders cannot be funded due to mismatches in
+// UTXO sizes, the orders that can be funded are funded. The values in the
+// MultiOrder parameter are funded in the order they are passed in.
+func (btc *baseWallet) FundMultiOrder(mo *asset.MultiOrder, maxLock uint64) ([]asset.Coins, [][]dex.Bytes, uint64, error) {
+	btc.log.Debugf("Attempting to fund a multi-order for %s, maxFeeRate = %d", btc.symbol, mo.MaxFeeRate)
+
+	var totalRequiredForOrders uint64
+	var swapInputSize uint64
+	if btc.segwit {
+		swapInputSize = dexbtc.RedeemP2WPKHInputTotalSize
+	} else {
+		swapInputSize = dexbtc.RedeemP2PKHInputSize
+	}
+	for _, value := range mo.Values {
+		if value.Value == 0 {
+			return nil, nil, 0, fmt.Errorf("cannot fund value = 0")
+		}
+		if value.MaxSwapCount == 0 {
+			return nil, nil, 0, fmt.Errorf("cannot fund zero-lot order")
+		}
+		req := calc.RequiredOrderFundsAlt(value.Value, swapInputSize, value.MaxSwapCount,
+			btc.initTxSizeBase, btc.initTxSize, mo.MaxFeeRate)
+		totalRequiredForOrders += req
+	}
+
+	if maxLock < totalRequiredForOrders && maxLock != 0 {
+		return nil, nil, 0, fmt.Errorf("maxLock < totalRequiredForOrders (%d < %d)", maxLock, totalRequiredForOrders)
+	}
+
+	bal, err := btc.Balance()
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("error getting wallet balance: %w", err)
+	}
+	if bal.Available < totalRequiredForOrders {
+		return nil, nil, 0, fmt.Errorf("insufficient funds. %d < %d",
+			bal.Available, totalRequiredForOrders)
+	}
+
+	if mo.FeeSuggestion > mo.MaxFeeRate {
+		return nil, nil, 0, fmt.Errorf("fee suggestion %d > max fee rate %d", mo.FeeSuggestion, mo.MaxFeeRate)
+	}
+	// Check wallets fee rate limit against server's max fee rate
+	if btc.feeRateLimit() < mo.MaxFeeRate {
+		return nil, nil, 0, fmt.Errorf(
+			"%v: server's max fee rate %v higher than configued fee rate limit %v",
+			dex.BipIDSymbol(BipID), mo.MaxFeeRate, btc.feeRateLimit())
+	}
+
+	customCfg := new(fundMultiOptions)
+	err = config.Unmapify(mo.Options, customCfg)
```

Comment:
I'd like if we can avoid marshalling into an in-memory ini file or using the go-ini package for this.  It very inefficient, and a large dep we should avoid getting dug in on further.

We've defined `splitKey` for instance, so I think we can do that for this too, as long as the docs for `fundMultiOptions` refer to the const keys so it is less likely to break with future changes.

Reply:
I've created a special `decodeFundMultiOptions` function. Is this what you were going for? Would it be better if we just passed json strings for all of these options instead of `map[string]string` and using ini?

Reply:
Yeah, that's good for now I think.  No silly ini intermediate.  Maybe about json strings, dunno.  It all comes back to the pain point of having all the json keys be strings too instead of typed, which would allow a direct unmarshal into a struct from json, but I recall the challenge is typing values on the frontend.  Good for now imo.

Reply:
> How is the end user going to set the values? Are they just expected to know what options are available for all assets?

Valid question.  `PreOrder` provides these options for regular trades, so perhaps it needs to work with multi-trades too (or a similar method) so that a bot frontend can show what options are available.

> The point of the INI stuff is so that we have some flexibility for assets to define their own options

IMO, INI only makes sense with reading a wallet setting file, not handling order options.   `OrderOptions` provided _to_ user/frontend  from preorder are a well-defined struct, while options returned _from_ frontend are `{"string": "string"}` objects that unmarshal naturally to `map[string]string`.  No ini use required there.

When we do `config.Unmapify` all over the place, we're just taking advantage of the `ini` tags on the struct fields.  This is what I find unfortunate.  There's no real reason we need to write to an in-memory ini-formatted buffer to convert `map[string]string` -> `struct` except because we decided to put the keys in the struct tags.  We could have instead written an unmarshal helper for each of these "options structs" and we'd not require the ini package or the intermediate data format.  I think we'll get bit by this.

> and the `map[string]string` is easy to store in the DB

True, it's easy, however `config.Data(options)` to prep bytes for a db bucket could just as easily have been JSON with little extra data overhead.

Reply:
> The point of the INI stuff is so that we have some flexibility for assets to define their own options and the map[string]string is easy to store in the DB, as opposed to say a custom struct. Are we going to store the user's bot programs to disk?

I just meant replacing the `ini` with json strings in the DB, just because @chappjc said `ini` was expensive. Not sure marshalling and unmarshalling json is much cheaper though.

> How is the end user going to set the values? Are they just expected to know what options are available for all assets?

The bots themselves have a configuration. The end-user only has to know the bot configuration, not the configuration for the wallets. On the UI we can hide the splitting related settings if neither of the wallets support it, but if someone is running market making using the RPC and a json config file, the settings would just be ignored if they are not relevant. For the UI, something like `PreOrder` would definitely be needed to know whether or not the settings should be hidden.